### PR TITLE
fix(query): auth jwt with ensure_user on metasrv

### PR DESCRIPTION
**Pull request by everpcpc**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #issue

### CODE REVIEW

File: src/query/service/src/auth.rs [hunk 1]
```diff
@@ -78,34 +80,55 @@ impl AuthMgr {
...
-                        for role in roles.clone().into_iter() {
-                            user_info.grants.grant_role(role);
+                let tenant = session.get_current_tenant();
+                let identity = UserIdentity::new(&user_name, "%");
+                let user = match user_api.get_user(&tenant, identity.clone()).await {
+                    Ok(user_info) => match user_info.auth_info {
+                        AuthInfo::JWT => Ok(user_info),
+                        _ => return Err(ErrorCode::AuthenticateFailure("wrong auth type")),
+                    },
+                    Err(e) => {
```
Comment (flaneur2020) [inline src/query/service/src/auth.rs:90]: how about short-circuit the logic in the front ?
if e.code() != ErrorCode::UNKNOWN_USER {
  return Err(ErrorCode::AuthenticateFailure(e.message()));
}

File: src/query/service/src/auth.rs [hunk 2]
```diff
@@ -78,34 +80,55 @@ impl AuthMgr {
...
+                        let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
+                        if user_info.identity().is_root() {
+                            return Err(ErrorCode::AuthenticateFailure(
+                                "root user is not allowed in jwt auth.",
+                            ));
+                        }
+                        if let Some(ref roles) = ensure_user.roles {
+                            for role in roles.clone().into_iter() {
+                                user_info.grants.grant_role(role);
+                            }
```
Comment (flaneur2020) [inline src/query/service/src/auth.rs:111]: in the earilier code, there's also a function call to initialize the builtin-roles here (including ACCOUNT_ADMIN and PUBLIC).

File: src/query/service/src/auth.rs [hunk 3]
```diff
@@ -78,34 +80,55 @@ impl AuthMgr {
...
+                            None => return Err(ErrorCode::AuthenticateFailure(e.message())),
+                        };
+                        // create a new user if not exists
+                        let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
+                        if user_info.identity().is_root() {
+                            return Err(ErrorCode::AuthenticateFailure(
+                                "root user is not allowed in jwt auth.",
+                            ));
+                        }
```
Comment (flaneur2020) [inline src/query/service/src/auth.rs:107]: I guess this check about `is_root()` do not needed?

with restricting the users with a JWT token must marked as JWT auth type, the root user's auth_type are less likely to be JWT, thus it also get filtered out.

File: src/query/service/src/auth.rs (hunk 2 above)
Comment (everpcpc) [reply src/query/service/src/auth.rs:111]: It was moved to the init phase.

**Review by flaneur2020 — APPROVED** [review]
LGTM :heart:

File: src/query/service/src/auth.rs [hunk 4]
```diff
@@ -78,34 +80,47 @@ impl AuthMgr {
...
+                        user_api.add_user(&tenant, user_info, true).await?;
+                        user_api.get_user(&tenant, identity).await
                     }
-                }
+                }?;
+
+                let user = match &user.auth_info {
+                    AuthInfo::JWT => user,
+                    _ => return Err(ErrorCode::AuthenticateFailure("wrong auth type")),
+                };
```
Comment (flaneur2020) [inline src/query/service/src/auth.rs:113]: I guess this check is duplicated with L86: https://github.com/datafuselabs/databend/pull/10786/files#diff-fe9c4d4d86fb6815af28d24728a7e98cc84952c53d2edd270a76bdd7279cd333R86

File: src/query/service/src/auth.rs [hunk 5]
```diff
@@ -78,34 +80,47 @@ impl AuthMgr {
...
+                            .ok_or(ErrorCode::AuthenticateFailure(e.message()))?;
+                        // create a new user if not exists
+                        let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
+                        if let Some(ref roles) = ensure_user.roles {
+                            for role in roles.clone().into_iter() {
+                                user_info.grants.grant_role(role);
+                            }
                         }
+                        user_api.add_user(&tenant, user_info, true).await?;
+                        user_api.get_user(&tenant, identity).await
```
Comment (flaneur2020) [inline src/query/service/src/auth.rs:106]: just returns `user_info` is ok? this may reduce an rpc call

Comment (BohuTANG) [reply src/query/service/src/auth.rs]: Mark resolved?

**Pull request merged**